src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -1091,6 +1099,29 @@ public void TrimExcess(int capacity)
             CopyEntries(oldEntries, oldCount);
         }
 
+        private void CopyEntriesSameLength(Dictionary<TKey, TValue> source)
+        {
+            Debug.Assert(_entries is not null);
+            Debug.Assert(_buckets is not null);
+            Debug.Assert(source is not null);
+            Debug.Assert(source._entries is not null);
+            Debug.Assert(source._buckets is not null);
+
+            int startVersion = source._version;
```

Comment:
To be completely safe this version should be retrieved in the calling function before `if (source._entries.Length == _entries.Length)` to know that at the end of this method that source._entries wasn't grown between the length comparison and the saving of source._version here. Let me know if we should save the version outside and possibly not use it if the length's don't match.